### ISSUE TITLE
Feature request: insert image

### ISSUE THREAD

**Issue by saknarak** [guigrpa/docx-templates#14]
It would be nice if it can insert image in loop
something like this

```
+++FOR staff in staffs+++
+++IMG $staff.photo+++
+++IMG barcode($staff.code)+++
+++IMG signature($staff.code)+++
+++END-FOR+++
```

```
createReport({
  template: 'templates/myTemplate.docx',
  output: 'reports/myReport.docx',
  data: {
    staffs: [
      {code: '123456', name: 'John', surname: 'Appleseed', photo: '/path/to/photo.jpg'},
    ],
  },
  fn: {
    barcode(code) {
      return Buffer(code).toString('base64') // base64 image of barcode
    },
    async signature(code) {
      const img = await axios.get('/path/to/signature/' + code, {responseType: 'arrayBuffer'}))
      return img.toString('base64')
    },
  },
});
```

Or any suggestions to implement this.

**Comment by guigrpa** [comment]
This is related to #3 (just for reference).

The idea of inserting images dynamically is in the roadmap, but nobody has been able to dedicate some time until now. Yesterday I pushed a commit that added the possibility to run async code in a JS snippet. This should pave the way for image loading later on.

**Comment by saknarak** [comment]
Thank you for update and great work.

**Comment by vdechef** [comment]
I am interested in this too: I will soon need this functionnality in my webapp.
I will try to find some time to work on this.

**Comment by guigrpa** [comment]
Don't worry, I'm taking this

**Comment by guigrpa** [comment]
Proposed syntax: `IMAGE <width-cm> <height-cm> <*>`. `*` would be a JS snippet returning a path string, a buffer, or a Promise of any of those. Paths would only be supported in Node.

You can do whatever you want in your JS snippet to return those, including calling functions. Functions can be defined in `fn`, as suggested by @saknarak.

For the moment, only *inline* images will be supported.

What do you think?

**Comment by guigrpa** [comment]
Wait, even better: `IMAGE <*>`. The JS snippet returns an object, or the Promise of an object.  The object would have this form:

```
```

**Comment by guigrpa** [comment]
Wait, even better: IMAGE <*>. The JS snippet returns an object, or the Promise of an object. The object would have this form:

```
{
  width: number,
  height: number,
  path?: string,
  buffer?: Buffer,
}
```

This way, we can obtain width and height dynamically (if we want), the syntax is simpler, and we can add more functionality in the future: maybe anchored positioning, effects, etc.

**Comment by vdechef** [comment]
So you mean that in the template, to generate a list of images with fixed width and dynamic height, we would have something like this ?

```
+++FOR img IN images+++
+++IMAGE 
{
  width: 12.1,
  height: img.height,
  path: img.path
}
+++
+++END-FOR $img+++
```

**Comment by guigrpa** [comment]
Exactly. Just notice you should use `$img` in your example, rather than `img`.

In other use cases, you may end up with something like: `+++IMAGE itemImage($item)+++`, and in your configuration:

```js
{
  fn: {
    itemImage: async (item) => {
      const imgBuffer = await fetchMyImage(item);
      // [... calculate width, height ...]
      return { width, height, buffer: imgBuffer };
    },
  },
}
```

**Comment by guigrpa** [comment]
We could also add functions to the `data` config, rather than a new abstract `fn` one.

**Comment by saknarak** [comment]
I had tested, it's work perfectly.
Thanks for your great work.

**Comment by guigrpa** [comment]
Have you already tested what's in master? In Node or in the browser?

**Comment by saknarak** [comment]
I tested with node

**Comment by guigrpa** [comment]
@vdechef Check out the webpack example. You can build and run it from `packages/example-webpack` with `yarn start`. It's 109 KB compressed.

**Comment by guigrpa** [comment]
Btw, to install dependencies just run `yarn` from the root folder.

**Comment by guigrpa** [comment]
Shipped in v2.4.0!

**Comment by jjhbw** [comment]
> I know this is a very old thread and images are now supported but is there any plans on development to add position parameters on images? 

Not sure what you mean exactly, but I'm not planning any more complex new features. Well-tested pull requests are welcome.
